traj/src/traj/fanuc_limits.py

```diff
@@ -0,0 +1,134 @@
+    Assumptions are: 20 elements per table, elements of type float, sorted in
+    descending order (ie: max limit -> fully scaled down limit).
+
+    Args:
+        no_load_thresh: threshold table for NO load configuration
+          list(float)
+        max_load_thresh: threshold table for MAX load configuration
+          list(float)
+        payload_max: maximum payload supported by the robot (Kg)
+          float
+        cart_vmax: maximum Cartesian velocity supported by the robot (mm/s)
+          default: 4e3
+          float
+        num_steps: number of entries in a single threshold table
+          default: 20
+          int
+        interp_func: order of interpolation used. Passed on to interp2d(..)
+          default: 'linear'
+          str
+
+    Returns:
+        Function wrapping the return value of scipy.interp2d(..).
+
+        Args:
+          cart_vel: the Cartesian velocity of the TCP (mm/s)
+            float
+          payload: the weight of the current payload of the robot (Kg)
+            default: payload_max
+            float
+
+        Returns:
+          2D interpolated joint limit for the given Cartesian velocity and
+          payload.
+
+    Example:
+
+      # create interpolation function for the acceleration limits of J1, with
+      # a maximum payload of 25 Kg, and the default maximum Cartesian velocity
+      # (of 4000 mm/s), default number of elements in the threshold tables (20)
+      # and the default interpolation strategy (linear).
+      j1_acc_limit_func = gen_limit_interpolation_func(
+          no_load_thresh=[2050.00, 2050.00, ..],
+          max_load_thresh=[1601.56, 1601.56, ..],
+          payload_max=25.0
+      )
+
+      # determine acceleration limit for J1 with TCP moving at 1.5 m/s and
+      # with a current payload of 6.3 Kg
+      j1_curr_acc_limit = j1_acc_limit_func(cart_vel=1.5e3, payload=6.3)[0]
+
+      # determine acceleration limits for J1 with TCP moving at 1100, 1350 and
+      # 1470 mm/s and the default (ie: max) payload
+      j1_acc_limits = j1_acc_limit_func(cart_vel=[1100, 1350, 1470])
+    """
+    len_nlt = len(no_load_thresh)
+    len_mlt = len(max_load_thresh)
+    if len_nlt != num_steps or len_mlt != num_steps:
+        raise ValueError(
+            "Threshold table should contain {} elements (got: {} and {} "
+            "elements for NO and MAX load respectively)"
+            .format(num_steps, len_nlt, len_mlt))
+
+    # TODO: check for negative max payloads
+    # TODO: check for negative max cart vel
+    # TODO: check for negative num steps
+
+    # TODO: this sets up a full 2D interpolation. Not sure that is what
+    # we want. Perhaps we do need to consider the 'binning' on the X-axis
+    # (ie: percentage of max cart velocity)
+    x = np.linspace(cart_vmax/num_steps, cart_vmax, num_steps)
+    y = [0.0, payload_max]
+    z = [no_load_thresh, max_load_thresh]
+    limit_interp2d = interp2d(x, y, z, kind=interp_func)
```

Comment:
Shouldn't Z have more than 2 elements? From the docstring for interp2d:
```
z : array_like
    The values of the function to interpolate at the data points. If
    `z` is a multi-dimensional array, it is flattened before use.  The
    length of a flattened `z` array is either
    len(`x`)*len(`y`) if `x` and `y` specify the column and row coordinates
    or ``len(z) == len(x) == len(y)`` if `x` and `y` specify coordinates
    for each point.
```
So z should have 20*2=40 elements...?

Reply:
> If `z` is a multi-dimensional array, it is flattened before use.

?

Reply:
Ah i see. I hadn't realized no_load_thresh and max_load_thresh were arrays. Makes sense now.